rest/database_init_manager.go

```diff
@@ -0,0 +1,312 @@
+			return doneChan, err
+		}
+		// For a mismatch in collections, stop and remove the existing worker, then continue through to creation of new worker
+		dbInitWorker.Stop()
+		delete(m.workers, dbConfig.Name)
+	}
+
+	base.InfofCtx(ctx, base.KeyAll, "Starting new async initialization for database %s ...",
+		base.MD(dbConfig.Name))
+	couchbaseCluster, err := CreateCouchbaseClusterFromStartupConfig(startupConfig, base.PerUseClusterConnections)
+	if err != nil {
+		return nil, err
+	}
+
+	bucketName := dbConfig.Name
+	if dbConfig.Bucket != nil {
+		bucketName = *dbConfig.Bucket
+	}
+
+	// Initialize ClusterN1QLStore for the bucket.  Scope and collection name are set
+	// per-operation
+	n1qlStore, err := couchbaseCluster.GetClusterN1QLStore(bucketName, "", "")
+	if err != nil {
+		return nil, err
+	}
+
+	indexOptions := m.BuildIndexOptions(startupConfig, dbConfig)
+
+	// Create new worker and add this caller as a watcher
+	worker := NewDatabaseInitWorker(ctx, dbConfig.Name, n1qlStore, collectionSet, indexOptions, m.collectionCompleteCallback)
+	m.workers[dbConfig.Name] = worker
+	doneChan, err = worker.addWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	// Start a goroutine to perform the initialization
+	go func() {
+		defer couchbaseCluster.Close()
+		// worker.Run blocks until completion, and returns any error on doneChan.
+		worker.Run()
+		if m.databaseCompleteCallback != nil {
+			m.databaseCompleteCallback(dbConfig.Name)
+		}
+		// On success, remove worker
+		m.workersLock.Lock()
+		delete(m.workers, dbConfig.Name)
+		m.workersLock.Unlock()
+	}()
+	return doneChan, nil
+}
+
+func (m *DatabaseInitManager) HasActiveInitialization(dbName string) bool {
+	m.workersLock.Lock()
+	defer m.workersLock.Unlock()
+	_, ok := m.workers[dbName]
+	return ok
+}
+
+func (m *DatabaseInitManager) BuildIndexOptions(startupConfig *StartupConfig, dbConfig *DatabaseConfig) db.InitializeIndexOptions {
+	numReplicas := DefaultNumIndexReplicas
+	if dbConfig.NumIndexReplicas != nil {
+		numReplicas = *dbConfig.NumIndexReplicas
+	}
+	return db.InitializeIndexOptions{
+		FailFast:    false,
+		NumReplicas: numReplicas,
+		Serverless:  startupConfig.IsServerless(),
+		UseXattrs:   dbConfig.UseXattrs(),
+	}
+}
+
+// Intended for test usage.  Updates to callback function aren't synchronized
+func (m *DatabaseInitManager) SetCallbacks(collectionComplete collectionCallbackFunc, databaseComplete func(dbName string)) {
+	m.collectionCompleteCallback = collectionComplete
+	m.databaseCompleteCallback = databaseComplete
+}
+
+func (m *DatabaseInitManager) Cancel(dbName string) {
+	m.workersLock.Lock()
+	defer m.workersLock.Unlock()
+	worker, ok := m.workers[dbName]
+	if !ok {
+		return
+	}
+	worker.Stop()
+}
+
+// buildCollectionIndexData determines the set of indexes required for each collection in the config, including
+// the metadata collection
+func (m *DatabaseInitManager) buildCollectionIndexData(config *DatabaseConfig) CollectionInitData {
```

Comment:
this is again duplicate code from server_context.go - i'm a bit less concerned about this vs index options.

We discussed offline about an enhancement on the main branch to make all the code use async pathway and just block on all workers being done.